qualtran/drawing/qpic_diagram.py

```diff
@@ -0,0 +1,332 @@
+
+
+def _wire_name_prefix_for_soq(soq: Soquet) -> str:
+    return soq.pretty()
+
+
+def _format_label_text(label: str, scale: Optional[float] = None) -> str:
+    replacements = {'&': r'\&', '_': r'\_', '<': r'$<$', '>': r'$>$', '⨁': r'$\oplus$'}
+    for key, val in replacements.items():
+        label = label.replace(key, val)
+    scale_prefix = r'\scalebox{' + str(scale) + r'}{' if scale else ''
+    scale_suffix = r'}' if scale else ''
+    return r'\textrm{' + scale_prefix + str(label) + scale_suffix + r'}'
+
+
+def _soq_symbol_to_text(symbol: WireSymbol) -> str:
+    assert hasattr(symbol, 'text')
+    symbol_text = symbol.text.strip()
+    if not symbol_text:
+        symbol_text = 'SJ'
+    return _format_label_text(symbol_text)
+
+
+def _soq_symbol_width(symbol: WireSymbol) -> int:
+    assert hasattr(symbol, 'text')
+    symbol_text = symbol.text.strip()
+    if not symbol_text:
+        symbol_text = 'SJ'
+    return 5 + 5 * len(symbol_text)
+
+
+class QpicWireManager:
+    """Methods to manage allocation/deallocation of wires for QPIC diagrams.
+
+    QPIC places wires in the order in which they are defined. For each soquet, the wire manager
+    allocates a new wire with prefix `_wire_name_prefix_for_soq(soq)` and an integer suffix that
+    corresponds to the smallest integer which does not correspond to an allocated wire.
+    """
+
+    def __init__(self):
+        self._soq_to_wire_name_tuple: Dict[Soquet, Tuple[str, int]] = {}
+        self._alloc_wires_with_prefix: Dict[str, Set[int]] = defaultdict(set)
+
+    def _wire_name_tuple_for_soq(self, soq: Soquet) -> Tuple[str, int]:
+        prefix = _wire_name_prefix_for_soq(soq)
+        allocated_suffixes = self._alloc_wires_with_prefix[prefix]
+        next_i = next(i for i in range(len(allocated_suffixes) + 1) if i not in allocated_suffixes)
+        return prefix, next_i
+
+    def _wire_name_tuple_to_str(self, wire_name: Tuple[str, int]) -> str:
+        prefix, i = wire_name
+        return prefix + '_' + str(i) if i else prefix
+
+    def alloc_wire_for_soq(self, soq: Soquet) -> str:
+        prefix, i = self._wire_name_tuple_for_soq(soq)
+        self._alloc_wires_with_prefix[prefix].add(i)
+        self._soq_to_wire_name_tuple[soq] = (prefix, i)
+        return self._wire_name_tuple_to_str((prefix, i))
+
+    def dealloc_wire_for_soq(self, soq: Soquet) -> str:
+        prefix, i = self._soq_to_wire_name_tuple[soq]
+        self._alloc_wires_with_prefix[prefix].remove(i)
+        self._soq_to_wire_name_tuple.pop(soq)
+        return self._wire_name_tuple_to_str((prefix, i))
+
+    def soq_to_wirename(self, soq) -> str:
+        assert soq in self._soq_to_wire_name_tuple
+        return self._wire_name_tuple_to_str(self._soq_to_wire_name_tuple[soq])
+
+    def soq_to_wirelabel(self, soq) -> str:
+        return _format_label_text(self.soq_to_wirename(soq))
+
+
+class QpicCircuit:
+    """Builds data corresponding to the input specification of a QPIC diagram"""
```

Comment:
can you document this class and its public methods or make this class or its methods private? document public attributes